core/embed/io/ble/stm32/ble.c

```diff
@@ -0,0 +1,505 @@
+  nrf_send_msg(NRF_SERVICE_BLE_MANAGER, &cmd, sizeof(cmd), NULL, NULL);
+
+  return true;
+}
+
+static void ble_send_pairing_reject(void) {
+  uint8_t cmd = INTERNAL_CMD_REJECT_PAIRING;
+  nrf_send_msg(NRF_SERVICE_BLE_MANAGER, &cmd, sizeof(cmd), NULL, NULL);
+}
+
+static void ble_send_pairing_accept(void) {
+  uint8_t cmd = INTERNAL_CMD_ALLOW_PAIRING;
+  nrf_send_msg(NRF_SERVICE_BLE_MANAGER, &cmd, sizeof(cmd), NULL, NULL);
+}
+
+static void ble_process_rx_msg_status(const uint8_t *data, uint32_t len) {
+  ble_driver_t *drv = &g_ble_driver;
+
+  if (!drv->initialized) {
+    return;
+  }
+  if (len < sizeof(event_status_msg_t)) {
+    // insufficient data length
+    return;
+  }
+
+  event_status_msg_t msg;
+  memcpy(&msg, data, sizeof(event_status_msg_t));
+
+  if (!drv->status_valid) {
+    if (msg.peer_count > 0) {
+      drv->mode_requested = BLE_MODE_CONNECTABLE;
+    } else {
+      drv->mode_requested = BLE_MODE_OFF;
+    }
+  }
+
+  if (drv->connected != msg.connected) {
+    if (msg.connected) {
+      // new connection
+
+      ble_event_t event = {.type = BLE_CONNECTED};
+      tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event),
+                      NULL);
+
+    } else {
+      // connection lost
+      ble_event_t event = {.type = BLE_DISCONNECTED};
+      tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event),
+                      NULL);
+
+      if (drv->mode_current == BLE_MODE_PAIRING) {
+        drv->mode_requested = BLE_MODE_CONNECTABLE;
+      }
+    }
+
+    drv->connected = msg.connected;
+  }
+
+  if (msg.advertising && !msg.advertising_whitelist) {
+    drv->mode_current = BLE_MODE_PAIRING;
+  } else if (msg.advertising) {
+    drv->mode_current = BLE_MODE_CONNECTABLE;
+  } else {
+    drv->mode_current = BLE_MODE_OFF;
+  }
+
+  drv->peer_count = msg.peer_count;
+
+  drv->status_valid = true;
+}
+
+static void ble_process_rx_msg_pairing_request(const uint8_t *data,
+                                               uint32_t len) {
+  ble_driver_t *drv = &g_ble_driver;
+  if (!drv->initialized) {
+    return;
+  }
+
+  if (len < 7) {
+    // insufficient data length
+    return;
+  }
+
+  if (drv->mode_requested != BLE_MODE_PAIRING ||
+      drv->mode_current != BLE_MODE_PAIRING) {
+    ble_send_pairing_reject();
+    return;
+  }
+
+  ble_event_t event = {.type = BLE_PAIRING_REQUEST, .data_len = 6};
+  memcpy(event.data, &data[1], 6);
+  tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event), NULL);
+}
+
+static void ble_process_rx_msg_pairing_cancelled(const uint8_t *data,
+                                                 uint32_t len) {
+  ble_driver_t *drv = &g_ble_driver;
+  if (!drv->initialized) {
+    return;
+  }
+
+  ble_event_t event = {.type = BLE_PAIRING_CANCELLED, .data_len = 0};
+  tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event), NULL);
+}
+
+static void ble_process_rx_msg(const uint8_t *data, uint32_t len) {
+  switch (data[0]) {
```

Comment:
Is it safe to assume that `len` > 0?

Reply:
probably mostly safe but i added the check anyway in https://github.com/trezor/trezor-firmware/commit/191fa1a026fb38d579b593ff0ccff0fe2f32bccd